Couldn't resolve value for subQuery in query with multiple references to same table

### CrateDB version

5.9.6 , 5.9.8 , and 5.10.0

### CrateDB setup information

Single node and 3 nodes

### Problem description

The query below fails with a `SQLParseException`

### Steps to Reproduce

```sql
CREATE TABLE hernan.places  (
	place_id INTEGER,
	place_name TEXT,
	parent_place_id INTEGER	
);

INSERT INTO hernan.places VALUES (1,'Vienna',2),(2,'Austria',null);

REFRESH TABLE hernan.places;

CREATE VIEW hernan.vw_places AS SELECT * FROM hernan.places;

CREATE TABLE hernan.tbl2 (field1 TEXT,place_id	 INTEGER);

INSERT INTO hernan.tbl2 VALUES ('abc',1);

REFRESH TABLE hernan.tbl2;

CREATE VIEW hernan.vw2 as select * from hernan.tbl2;

WITH cities AS (
	SELECT g.place_id as city_id,g.parent_place_id,g.place_name as city_name
	FROM hernan.vw_places g
	WHERE g.parent_place_id=2
),
countries AS (
	SELECT s.place_name,s.place_id
	FROM hernan.vw_places s
	WHERE place_id IN (SELECT parent_place_id FROM cities)
),
cities_with_countries AS (
	SELECT *,
		(SELECT countries.place_name
		FROM countries
		WHERE cities.parent_place_id=countries.place_id) as country_name
	FROM cities
),b as (
        SELECT field1,
            place_id
        FROM hernan.vw2
        WHERE  place_id IN (SELECT city_id FROM cities)
            
    )
SELECT *
FROM b
INNER JOIN cities_with_countries c ON c.city_id = b.place_id;
```

### Actual Result

```
io.crate.exceptions.SQLParseException: Couldn't create execution plan from logical plan because of: Couldn't resolve value for subQuery: (SELECT parent_place_id FROM (cities)):
Eval[place_name] (rows=0)
  └ Rename[place_name, place_id] AS countries (rows=0)
    └ Limit[2::bigint;0::bigint] (rows=0)
      └ Filter[(place_id = parent_place_id)] (rows=0)
        └ MultiPhase (rows=unknown)
          └ Rename[place_name, place_id] AS s (rows=unknown)
            └ Rename[place_name, place_id] AS hernan.vw_places (rows=unknown)
              └ Collect[hernan.places | [place_name, place_id] | (place_id = ANY((SELECT parent_place_id FROM (cities))))] (rows=unknown)
          └ Rename[parent_place_id] AS cities (rows=unknown)
            └ Rename[parent_place_id] AS g (rows=unknown)
              └ Rename[parent_place_id] AS hernan.vw_places (rows=unknown)
                └ OrderBy[parent_place_id ASC] (rows=unknown)
                  └ Collect[hernan.places | [parent_place_id] | (parent_place_id = 2)] (rows=unknown)
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:211)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:200)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:176)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:130)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2238)
	at io.crate.rest.action.RestResultSetReceiver.fail(RestResultSetReceiver.java:92)
	at io.crate.session.RetryOnFailureResultReceiver.fail(RetryOnFailureResultReceiver.java:100)
	at io.crate.session.RowConsumerToResultReceiver.accept(RowConsumerToResultReceiver.java:67)
	at io.crate.execution.engine.InterceptingRowConsumer.tryForwardResult(InterceptingRowConsumer.java:95)
	at io.crate.execution.engine.InterceptingRowConsumer.accept(InterceptingRowConsumer.java:75)
	at io.crate.execution.jobs.CumulativePageBucketReceiver.kill(CumulativePageBucketReceiver.java:297)
	at io.crate.execution.jobs.DistResultRXTask.kill(DistResultRXTask.java:71)
	at io.crate.execution.jobs.RootTask$TaskFinishedListener.onFailure(RootTask.java:379)
	at io.crate.execution.jobs.RootTask$TaskFinishedListener.accept(RootTask.java:392)
	at io.crate.execution.jobs.RootTask$TaskFinishedListener.accept(RootTask.java:335)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2238)
	at io.crate.execution.jobs.CumulativePageBucketReceiver.lambda$new$1(CumulativePageBucketReceiver.java:115)
	at io.crate.execution.engine.distribution.merge.BatchPagingIterator.close(BatchPagingIterator.java:99)
	at io.crate.execution.engine.distribution.merge.BatchPagingIterator.kill(BatchPagingIterator.java:169)
	at io.crate.execution.jobs.CumulativePageBucketReceiver.kill(CumulativePageBucketReceiver.java:288)
	at io.crate.execution.engine.distribution.TransportDistributedResultAction.nodeOperation(TransportDistributedResultAction.java:183)
	at io.crate.execution.engine.distribution.TransportDistributedResultAction.nodeOperation(TransportDistributedResultAction.java:133)
	at io.crate.execution.support.NodeActionRequestHandler.messageReceived(NodeActionRequestHandler.java:48)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:59)
	at org.elasticsearch.transport.TransportService.sendLocalRequest(TransportService.java:677)
	at org.elasticsearch.transport.TransportService$2.sendRequest(TransportService.java:109)
	at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:617)
	at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:535)
	at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:487)
	at io.crate.execution.support.Transports.sendRequest(Transports.java:68)
	at io.crate.execution.support.Transports.sendRequest(Transports.java:80)
	at io.crate.execution.engine.distribution.TransportDistributedResultAction.doExecute(TransportDistributedResultAction.java:123)
	at io.crate.execution.engine.distribution.TransportDistributedResultAction.doExecute(TransportDistributedResultAction.java:65)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:46)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:40)
	at org.elasticsearch.client.node.NodeClient.execute(NodeClient.java:57)
	at io.crate.execution.engine.distribution.DistributingConsumerFactory.lambda$new$0(DistributingConsumerFactory.java:60)
	at io.crate.execution.engine.distribution.DistributingConsumer.forwardFailure(DistributingConsumer.java:156)
	at io.crate.execution.engine.distribution.DistributingConsumer.lambda$consumeIt$0(DistributingConsumer.java:133)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2238)
	at io.crate.data.CollectingRowConsumer.accept(CollectingRowConsumer.java:45)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:601)
	at io.crate.planner.operators.LogicalPlanner.lambda$execute$4(LogicalPlanner.java:574)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:907)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:885)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:554)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2223)
	at io.crate.data.BatchIterators.lambda$collect$1(BatchIterators.java:50)
	at io.crate.data.BatchIterator.move(BatchIterator.java:167)
	at io.crate.data.BatchIterators.collect(BatchIterators.java:47)
	at io.crate.data.CollectingRowConsumer.accept(CollectingRowConsumer.java:40)
	at io.crate.execution.engine.InterceptingRowConsumer.lambda$tryForwardResult$1(InterceptingRowConsumer.java:93)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: java.lang.IllegalArgumentException: Couldn't create execution plan from logical plan because of: Couldn't resolve value for subQuery: (SELECT parent_place_id FROM (cities)):
Eval[place_name] (rows=0)
  └ Rename[place_name, place_id] AS countries (rows=0)
    └ Limit[2::bigint;0::bigint] (rows=0)
      └ Filter[(place_id = parent_place_id)] (rows=0)
        └ MultiPhase (rows=unknown)
          └ Rename[place_name, place_id] AS s (rows=unknown)
            └ Rename[place_name, place_id] AS hernan.vw_places (rows=unknown)
              └ Collect[hernan.places | [place_name, place_id] | (place_id = ANY((SELECT parent_place_id FROM (cities))))] (rows=unknown)
          └ Rename[parent_place_id] AS cities (rows=unknown)
            └ Rename[parent_place_id] AS g (rows=unknown)
              └ Rename[parent_place_id] AS hernan.vw_places (rows=unknown)
                └ OrderBy[parent_place_id ASC] (rows=unknown)
                  └ Collect[hernan.places | [parent_place_id] | (parent_place_id = 2)] (rows=unknown)
	at io.crate.planner.operators.SubQueryResults.getSafe(SubQueryResults.java:62)
	at io.crate.planner.operators.SubQueryAndParamBinder.visitSelectSymbol(SubQueryAndParamBinder.java:70)
	at io.crate.planner.operators.SubQueryAndParamBinder.visitSelectSymbol(SubQueryAndParamBinder.java:37)
	at io.crate.expression.symbol.SelectSymbol.accept(SelectSymbol.java:94)
	at io.crate.expression.symbol.FunctionCopyVisitor.twoArgs(FunctionCopyVisitor.java:97)
	at io.crate.expression.symbol.FunctionCopyVisitor.processAndMaybeCopy(FunctionCopyVisitor.java:57)
	at io.crate.expression.symbol.FunctionCopyVisitor.visitFunction(FunctionCopyVisitor.java:154)
	at io.crate.expression.symbol.FunctionCopyVisitor.visitFunction(FunctionCopyVisitor.java:40)
	at io.crate.expression.symbol.Function.accept(Function.java:251)
	at io.crate.planner.operators.SubQueryAndParamBinder.apply(SubQueryAndParamBinder.java:83)
	at io.crate.planner.operators.SubQueryAndParamBinder.apply(SubQueryAndParamBinder.java:37)
	at java.base/java.util.function.Function.lambda$andThen$1(Function.java:88)
	at io.crate.analyze.WhereClause.map(WhereClause.java:188)
	at io.crate.planner.operators.Collect.createPhase(Collect.java:243)
	at io.crate.planner.operators.Collect.build(Collect.java:145)
	at io.crate.planner.operators.Rename.build(Rename.java:193)
	at io.crate.planner.operators.Rename.build(Rename.java:193)
	at io.crate.planner.operators.MultiPhase.build(MultiPhase.java:73)
	at io.crate.planner.operators.Filter.build(Filter.java:85)
	at io.crate.planner.operators.Limit.build(Limit.java:111)
	at io.crate.planner.operators.Rename.build(Rename.java:193)
	at io.crate.planner.operators.Eval.build(Eval.java:83)
	at io.crate.planner.operators.RootRelationBoundary.build(RootRelationBoundary.java:60)
	at io.crate.planner.operators.LogicalPlanner.getNodeOperationTree(LogicalPlanner.java:620)
	at io.crate.planner.operators.LogicalPlanner.doExecute(LogicalPlanner.java:599)
	... 13 more
Caused by: java.lang.IllegalArgumentException: Couldn't resolve value for subQuery: (SELECT parent_place_id FROM (cities))
	... 38 more
```

### Expected Result

```
 field1 | place_id | city_id | parent_place_id | city_name | country_name 
--------+----------+---------+-----------------+-----------+--------------
 abc    |        1 |       1 |               2 | Vienna    | Austria
(1 row)
```

--- a/docs/appendices/release-notes/5.10.1.rst
+++ b/docs/appendices/release-notes/5.10.1.rst
@@ -74,3 +74,17 @@ Fixes
   ``ORDER BY`` and query had a ``LIMIT`` clause. Example.::
 
     SELECT a, b + ? as sum FROM tbl ORDER BY sum LIMIT 10
+
+- Fixed an issue that caused an error when using a correlated sub-query and
+  a normal subquery on the same level together with an additional sub-query on
+  an outer level. Example::
+
+    WITH tbl AS (SELECT 1 as x)
+    SELECT (
+      SELECT x FROM tbl
+      WHERE t.x = tbl.x
+        AND tbl.x IN (SELECT x FROM unnest([1]))
+    ),
+    (SELECT y FROM unnest([2]) as t2(y))
+    FROM tbl t
+

--- a/server/src/main/java/io/crate/planner/operators/SubQueryResults.java
+++ b/server/src/main/java/io/crate/planner/operators/SubQueryResults.java
@@ -22,6 +22,7 @@
 package io.crate.planner.operators;
 
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
@@ -46,11 +47,10 @@ public class SubQueryResults {
     private Row inputRow = Row.EMPTY;
 
     public SubQueryResults(Map<SelectSymbol, Object> valuesBySubQuery) {
-        this.valuesBySubQuery = valuesBySubQuery;
-        this.boundOuterColumns = EMPTY_OUTER_COLUMNS;
+        this(valuesBySubQuery, EMPTY_OUTER_COLUMNS);
     }
 
-    public SubQueryResults(Map<SelectSymbol, Object> valuesBySubQuery,
+    private SubQueryResults(Map<SelectSymbol, Object> valuesBySubQuery,
                            ObjectIntMap<OuterColumn> boundOuterColumns) {
         this.valuesBySubQuery = valuesBySubQuery;
         this.boundOuterColumns = boundOuterColumns;
@@ -90,7 +90,15 @@ public class SubQueryResults {
     }
 
     public static SubQueryResults merge(SubQueryResults subQueryResults1, SubQueryResults subQueryResults2) {
-        // Correlated subquery related fields.
+        // Correlated subquery related fields, expect only one of them to contain correlated subquery fields.
+        assert subQueryResults1.boundOuterColumns.isEmpty()
+            || subQueryResults2.boundOuterColumns.isEmpty()
+            : "Can't merge two SubQueryResults with non-empty boundOuterColumns";
+
+        assert subQueryResults1.inputRow.equals(Row.EMPTY)
+            || subQueryResults2.inputRow.equals(Row.EMPTY)
+            : "Can't merge two SubQueryResults with non-empty inputRow";
+
         Row mergedRow = subQueryResults1.inputRow;
         ObjectIntMap<OuterColumn> mergedBoundOuterColumns = subQueryResults1.boundOuterColumns;
         if (mergedRow.equals(Row.EMPTY)) {
@@ -100,11 +108,11 @@ public class SubQueryResults {
             mergedBoundOuterColumns = subQueryResults2.boundOuterColumns;
         }
 
-        // Regular sub-select field.
-        Map<SelectSymbol, Object> mergedValuesBySubQuery = subQueryResults1.valuesBySubQuery;
-        if (mergedValuesBySubQuery.isEmpty()) {
-            mergedValuesBySubQuery = subQueryResults2.valuesBySubQuery;
-        }
+        // Fully merge regular sub-select field.
+        Map<SelectSymbol, Object> mergedValuesBySubQuery = new HashMap<>();
+        mergedValuesBySubQuery.putAll(subQueryResults1.valuesBySubQuery);
+        mergedValuesBySubQuery.putAll(subQueryResults2.valuesBySubQuery);
+
 
         SubQueryResults combinedSubQueryResults = new SubQueryResults(mergedValuesBySubQuery, mergedBoundOuterColumns);
         combinedSubQueryResults.bindOuterColumnInputRow(mergedRow);

--- a/server/src/test/java/io/crate/integrationtests/CorrelatedSubqueryITest.java
+++ b/server/src/test/java/io/crate/integrationtests/CorrelatedSubqueryITest.java
@@ -446,7 +446,7 @@ public class CorrelatedSubqueryITest extends IntegTestCase {
      * Tests a bug https://github.com/crate/crate/issues/15398.
      */
     @Test
-    public void test_can_mix_correlated_qubquery_and_sub_select() {
+    public void test_can_mix_correlated_subquery_and_sub_select() {
         execute("CREATE TABLE tbl(x int)");
         execute("INSERT INTO tbl(x) VALUES (1)");
         execute("refresh table tbl");
@@ -462,6 +462,23 @@ public class CorrelatedSubqueryITest extends IntegTestCase {
         assertThat(response).hasRows("1");
     }
 
+    /*
+     * https://github.com/crate/crate/issues/17307
+     */
+    @Test
+    public void test_can_mix_correlated_subquery_and_sub_selects_with_sub_select() {
+        execute("""
+            WITH tbl AS (SELECT 1 as x)
+            SELECT (
+              SELECT x FROM tbl
+              WHERE t.x = tbl.x
+                AND
+                tbl.x IN (SELECT x FROM unnest([1]))
+            ), (SELECT y FROM unnest([2]) as t2(y))
+            FROM tbl t
+            """);
+        assertThat(response).hasRows("1| 2");
+    }
 
     /*
      * https://github.com/crate/crate/issues/16124

--- a/server/src/test/java/io/crate/planner/operators/SubQueryResultsTest.java
+++ b/server/src/test/java/io/crate/planner/operators/SubQueryResultsTest.java
@@ -1,0 +1,58 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.operators;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Map;
+
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Test;
+
+import io.crate.expression.symbol.SelectSymbol;
+import io.crate.testing.DummyRelation;
+import io.crate.types.DataTypes;
+
+public class SubQueryResultsTest extends ESTestCase {
+
+    @Test
+    public void test_merges_values_of_both_sides() {
+        SelectSymbol selectSymbol1 = new SelectSymbol(
+            new DummyRelation("x"),
+            DataTypes.STRING_ARRAY,
+            SelectSymbol.ResultType.SINGLE_COLUMN_SINGLE_VALUE,
+            false);
+        SelectSymbol selectSymbol2 = new SelectSymbol(
+            new DummyRelation("y"),
+            DataTypes.STRING_ARRAY,
+            SelectSymbol.ResultType.SINGLE_COLUMN_SINGLE_VALUE,
+            false);
+
+        SubQueryResults subQueryResults1 = new SubQueryResults(Map.of(selectSymbol1, "foo"));
+        SubQueryResults subQueryResults2 = new SubQueryResults(Map.of(selectSymbol2, "bar"));
+
+        SubQueryResults merged = SubQueryResults.merge(subQueryResults1, subQueryResults2);
+
+        assertThat(merged.getSafe(selectSymbol1)).isEqualTo("foo");
+        assertThat(merged.getSafe(selectSymbol2)).isEqualTo("bar");
+    }
+}

